---
title: "RiboSeq_Report"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_document
params:
    args: myarg
---
The following document is a report on the alignment of multiple Ribo-Seq datasets against NMD and RI transcripts as well as against the human genome. These alignments are then analyzed in respect to their mapping against the previously determined unique Regions of the also previously determined split-ORFs.

## Alignment against NMD transcripts

```{r, echo=FALSE,results = 'asis'}
library('knitr')
setwd(params[[1]][[1]])
#setwd("E:/Justin_Backup_29.04.2020/Justin/Uni-Frankfurt/FP+Masterarbeit/PipeTest/Pipeline/Output/run_11.02.2021-13.21.38_new_NMD/BOWTIE")
directories<-list.dirs(path = ".", full.names = TRUE, recursive = TRUE)

randomnmdfiles=list()
for(i in directories){
  if(!(identical(list.files(i,pattern="*NMD_random_intersect_counts_relative_sorted.bed"), character(0)))){
    randomnmdfiles=c(randomnmdfiles,paste0(i,"/",list.files(i,pattern="*NMD_random_intersect_counts_relative_sorted.bed")))
  }
}
randomnmddataframes <- lapply(randomnmdfiles, read.csv,header = FALSE,sep="\t")
backgroundnmd=list()
for(randomframe in randomnmddataframes){
  colnames(randomframe)=c("ID","start","stop","read_count", "relative_count")
  temp=t.test(randomframe$relative_count)[[4]][[2]]
  backgroundnmd = c(backgroundnmd,temp)
}
randomSetnames=list()
for(i in directories){
  if(!(identical(list.files(i,pattern="*NMD_random_intersect_counts_relative_sorted.bed"), character(0)))){
    randomSetnames=c(randomSetnames,list.files(i,pattern="*NMD_random_intersect_counts_relative_sorted.bed"))
  }
}
randomnames=stringr::str_replace(randomSetnames, pattern = "_random_intersect_counts_relative_sorted.bed", replacement = "")
printbackground=data.frame(x=backgroundnmd)
printbackground=as.data.frame(t(printbackground))
rownames(printbackground)=randomnames
colnames(printbackground)="Threshold"
print(kable(printbackground, caption="Thresholds for each dataset"))
threshold=backgroundnmd
#threshold=0
#for(k in backgroundnmd){
#  threshold = threshold + k
#}
#threshold=round(threshold/length(backgroundnmd), digits = 3)
#cat(paste0("The threshold for relevant regions, determined by random region coverage is: ", threshold))
```

```{r, echo=FALSE,results='asis'}
setwd(params[[1]][[1]])
#setwd("E:/Justin_Backup_29.04.2020/Justin/Uni-Frankfurt/FP+Masterarbeit/PipeTest/Pipeline/Output/run_11.02.2021-13.21.38_new_NMD/BOWTIE")
directories=list.dirs(path = ".", full.names = TRUE, recursive = TRUE)
nmdfiles=list()
setnames=list()
for(i in directories){
  if(!(identical(list.files(i,pattern="*NMD_intersect_counts_relative_sorted.bed"), character(0)))){
    nmdfiles=c(nmdfiles,paste0(i,"/",list.files(i,pattern="*NMD_intersect_counts_relative_sorted.bed")))
    setnames=c(setnames,list.files(i,pattern="*NMD_intersect_counts_relative_sorted.bed"))
  }
}
nmddataframes <- lapply(nmdfiles, read.csv,header = FALSE,sep="\t")
nmdlist=list()
for(a in nmddataframes){
  a$V1=paste0(a$V1,":",a$V4)
  a$V4=NULL
  nmdlist=c(nmdlist,list(a))
}
nmddataframes=nmdlist
names(nmddataframes) <- stringr::str_replace(setnames, pattern = "_intersect_counts_relative_sorted.bed", replacement = "")
relevantregionscount=list()
relevantregions=list()
i=1
for(frame in nmddataframes){
  colnames(frame)=c("ID","start","stop","read_count", "relative_count")
  temp=0
  c=1
  for(count in frame$relative_count){
    if(count>=threshold[[i]]){
      if(frame$read_count[[c]] > 1){
        temp = temp + 1
      }
    }
    c = c + 1
  }
  i = i + 1
  relevantregionscount = c(relevantregionscount,temp)
}
printframe=data.frame(x=relevantregionscount)
printframe=as.data.frame(t(printframe))
rownames(printframe)=names(nmddataframes)
colnames(printframe)=paste0("Number of unique regions with relative count >= threshold")
print(kable(printframe, caption="Regions above the threshold"))
#print(printframe)

```
The following table shows the top five unique regions with a relative count above the threshold (if available).

```{r, echo=FALSE,message=FALSE,results='asis'}
library("dplyr")
library("UpSetR")
library("knitr")
upsetlist=list()
j=1
for(f in nmddataframes){
  bed=f
  list<- rep(NA, length(bed[,1]))
  for(i in 1:length(bed[,1])){
    list[i]=paste0(bed[i,1],":",bed[i,2],":",bed[i,3])
  }
  rownames(bed)=list
  bed = filter(bed, V6 > threshold[[j]])
  bed = filter(bed, V5 > 1)
  upsetlist=c(upsetlist,list(row.names(bed)))
  colnames(bed)=c("ID","start","stop","read_count", "relative_count")
  #print(names(nmddataframes)[j])
  #print(bed[1:5,], row.names = FALSE)
  print(kable(bed[1:5,], caption=names(nmddataframes)[j], row.names = FALSE))
  j = j + 1
}
names(upsetlist)=names(nmddataframes)
```

The following plot shows how many unique regions are expressed in which dataset and how those overlap with the other data sets.

```{r, echo=FALSE,message=FALSE}
upset(fromList(upsetlist), order.by = "freq", nsets = 10, point.size = 2.25, line.size = 1.5, set_size.show = TRUE,
      set_size.scale_max = 1000, 
      mainbar.y.label = "Unique region Intersections", sets.x.label = "Matching unique regions per dataset",
      text.scale = c(1, 1, 0.9, 1, 1, 1))# c(1.5, 1.25, 0.9, 1.25, 1.25, 1.5)

```


## Alignment against RI transcripts

```{r, echo=FALSE,results = 'asis'}
setwd(params[[1]][[2]])
#setwd("E:/Justin_Backup_29.04.2020/Justin/Uni-Frankfurt/FP+Masterarbeit/PipeTest/Pipeline/Output/run_11.02.2021-15.10.13_new_RI/BOWTIE")
ridirectories=list.dirs(path = ".", full.names = TRUE, recursive = TRUE)

randomrifiles=list()
for(i in ridirectories){
  if(!(identical(list.files(i,pattern="*RI_random_intersect_counts_relative_sorted.bed"), character(0)))){
    randomrifiles=c(randomrifiles,paste0(i,"/",list.files(i,pattern="*RI_random_intersect_counts_relative_sorted.bed")))
  }
}
randomridataframes <- lapply(randomrifiles, read.csv,header = FALSE,sep="\t")
backgroundri=list()
for(randomriframe in randomridataframes){
  colnames(randomriframe)=c("ID","start","stop","read_count", "relative_count")
  temp2=t.test(randomriframe$relative_count)[[4]][[2]]
  backgroundri = c(backgroundri,temp2)
}
randomRiSetnames=list()
for(i in ridirectories){
  if(!(identical(list.files(i,pattern="*RI_random_intersect_counts_relative_sorted.bed"), character(0)))){
    randomRiSetnames=c(randomRiSetnames,list.files(i,pattern="*RI_random_intersect_counts_relative_sorted.bed"))
  }
}
randomRinames=stringr::str_replace(randomRiSetnames, pattern = "_random_intersect_counts_relative_sorted.bed", replacement = "")
printbackgroundRI=data.frame(x=backgroundri)
printbackgroundRI=as.data.frame(t(printbackgroundRI))
rownames(printbackgroundRI)=randomRinames
colnames(printbackgroundRI)="Threshold"
print(kable(printbackgroundRI, caption="Thresholds for each dataset"))
thresholdri=backgroundri
#thresholdri=0
#for(k in backgroundri){
#  thresholdri = thresholdri + k
#}
#thresholdri=round(thresholdri/length(backgroundri), digits = 3)
#cat(paste0("The threshold for relevant regions, determined by random region coverage is: ", thresholdri))
```

```{r, echo=FALSE,results='asis'}
setwd(params[[1]][[2]])
#setwd("E:/Justin_Backup_29.04.2020/Justin/Uni-Frankfurt/FP+Masterarbeit/PipeTest/Pipeline/Output/run_11.02.2021-15.10.13_new_RI/BOWTIE")
ridirectories=list.dirs(path = ".", full.names = TRUE, recursive = TRUE)
rifiles=list()
risetnames=list()
for(i in ridirectories){
  if(!(identical(list.files(i,pattern="*RI_intersect_counts_relative_sorted.bed"), character(0)))){
    rifiles=c(rifiles,paste0(i,"/",list.files(i,pattern="*RI_intersect_counts_relative_sorted.bed")))
    risetnames=c(risetnames,list.files(i,pattern="*RI_intersect_counts_relative_sorted.bed"))
  }
}
ridataframes <- lapply(rifiles, read.csv,header = FALSE,sep="\t")
rilist=list()
for(a in ridataframes){
  a$V1=paste0(a$V1,":",a$V4)
  a$V4=NULL
  rilist=c(rilist,list(a))
}
ridataframes=rilist
names(ridataframes) <- stringr::str_replace(risetnames, pattern = "_intersect_counts_relative_sorted.bed", replacement = "")
rirelevantregionscount=list()
rirelevantregions=list()
i=1
for(riframe in ridataframes){
  colnames(riframe)=c("ID","start","stop","read_count", "relative_count")
  temp2=0
  c=1
  for(count2 in riframe$relative_count){
    if(count2>=thresholdri[[i]]){
      if(riframe$read_count[[c]] > 1){
        temp2 = temp2 + 1
      }
    }
    c = c + 1
  }
  i = i + 1
  rirelevantregionscount = c(rirelevantregionscount,temp2)
}
riprintframe=data.frame(x=rirelevantregionscount)
riprintframe=as.data.frame(t(riprintframe))
rownames(riprintframe)=names(ridataframes)
colnames(riprintframe)=paste0("Number of unique regions with relative count >= Threshold")
print(kable(riprintframe, caption="Regions above the threshold"))
#print(riprintframe)

```
The following table shows the top five unique regions with a relative count above the threshold (if available).

```{r, echo=FALSE,message=FALSE,results='asis'}
library("dplyr")
library("UpSetR")
riupsetlist=list()
j=1
for(f in ridataframes){
  bed=f
  list<- rep(NA, length(bed[,1]))
  for(i in 1:length(bed[,1])){
    list[i]=paste0(bed[i,1],":",bed[i,2],":",bed[i,3])
  }
  rownames(bed)=list
  bed = filter(bed, V6 > thresholdri[[j]])
  bed = filter(bed, V5 > 1)
  riupsetlist=c(riupsetlist,list(row.names(bed)))
  colnames(bed)=c("ID","start","stop","read_count", "relative_count")
  #print(names(ridataframes)[j])
  #print(bed[1:5,], row.names = FALSE)
  print(kable(bed[1:5,], caption=names(ridataframes)[j], row.names = FALSE))
  j = j + 1
}
names(riupsetlist)=names(ridataframes)
```

The following plot shows how many unique regions are expressed in which dataset and how those overlap with the other data sets.

```{r, echo=FALSE,message=FALSE}
upset(fromList(riupsetlist), order.by = "freq", nsets = 10, point.size = 2.25, line.size = 1.5, set_size.show = TRUE,
      set_size.scale_max = 1000,
      mainbar.y.label = "Unique region Intersections", sets.x.label = "Matching unique regions per dataset",
      text.scale = c(1, 1, 0.9, 1, 1, 1)) #c(1.5, 1.25, 0.9, 1.25, 1.25, 1.5)

```