---
title: "Split-ORF Report"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: html_document
params:
    args: myarg
---

This report shows a number of summary statistics of the a result file obtained from the [SplitOrfs](https://github.com/SchulzLab/SplitOrfs) software.


## Overview of the valid alignment results:


```{r basic stats,echo=FALSE}
library(seqinr)
datapath=getwd()
path = "FinalAnalysis"
file1=paste0(datapath,params[[1]][[1]])
file2=paste0(datapath,params[[1]][[2]])
file3=paste0(datapath,params[[1]][[3]])

transcripts = read.fasta(file3)

#read in all the Ensembl IDs with transcript support level annotation
validIds=read.table(file1,header=T)#,fill=T,stringsAsFactors=F) 

tab = read.table(file2,header=T)#,fill=T,stringsAsFactors=F)

tab= subset(tab,tab$OrfTransID %in% validIds$OrfTransID)

NumTrans = length(transcripts)#nrow(validIds)
validTrans = length(unique(tab$OrfTransID))
annots=nrow(tab)-sum(is.na(tab$ORF.DomainAnnot))
uniqGenes = length(unique(tab$geneID))
#create plotting dataframe
legends=c("Number of transcripts",
          "Transcripts with >= 2 ORFs",
          "Transcripts with >= 1 PFAM domain overlap",
          "Number of unique Genes")

values=c(NumTrans,validTrans,annots,uniqGenes)
plotData = data.frame(Entry=legends,Values=values)
plotData
```

## Histogramm of matching ORFs per transcripts:

```{r hist1, echo=FALSE}
print(tab$NumOrfs)
hist(tab$NumOrfs,col="blue",breaks=100,main="Matching ORFs per transcript",xlab="Number of ORFs")
```

## Histogramm of NMD-protein pairs

```{r hist2, echo=FALSE}
hist(table(tab$geneID),breaks=100,col="blue",main="Number of transcript - protein pairs per gene",xlab="Transcript-protein pairs per gene")
```

## Top 20 genes, that have the most transcripts with >=2 ORFs
``` {r top pairs,echo=FALSE}
    df=data.frame(table(tab$geneID))
    names(df)=c("geneIDs","frequency")
    head(df[order(df$frequency,decreasing=T),],20)
    
```

## Histogramm of min and max sequence identity

A histogramm of min and max sequence identity values from the ORF-protein alignments, i.e., each such value is derived from one ORF-protein alignment. For all ORFs of a protein-ORF match the min respectively max seq. identity is recorded. The minimal sequence identity allowed for an alignment by default is 95%.

```{r hist3, echo=FALSE}
par(mfrow=c(2,1))
hist(tab$MinSeqIdent,col="blue",main="Min seq identities",xlab="Sequence identity value")
hist(tab$MaxSeqIdent,col="blue",main="Max seq identities",xlab="Sequence identity value")

```

## Histogram in the intervall [0,2000]

To analyze the length of the predicted orfs in DNA basepairs we plot a histogram in the intervall [0,2000]. Note that the minimal length of an ORF is 120 base pairs, which explains the sudden break close to 0. 

```{r orfLengths, echo=FALSE}
par(mfrow=c(2,1))
hist(as.numeric(unlist(strsplit(as.character(tab$OrfLengths),",",fixed=T))),xlim=c(0,2000), breaks=500,col="blue",xlab="Orf length",main="Nucleotide lengths of all Orfs")

hist(sapply(strsplit(as.character(tab$OrfLengths),",",fixed=T),function(x){return(mean(as.numeric(x)))}),xlim=c(0,800), breaks=300,col="blue",main="Mean orf lengths per gene",xlab="Mean nucleotide orf length")

```

## PFAM domain annotations

PFAM domain annotations were obtained for all protein coding genes. Here we show how often more than one ORF overlaps an annotated PFAM domain in the alignment region. For example a value of two means, that two distinct ORFs made from the transcript aligned to regions in the protein that overlapped PFAM annotations.

```{r orfAnnots, echo=FALSE}

hist(subset(tab,AnnotPercent >0)$NumORFAnnot,breaks=100,col="blue",main="Frequency of ORFs with >= 1 annotation",xlab="Number of ORFs that are overlapping annotation")

```

## Top 20 genes that have the most transcripts ORFs with annotations per transcript

These are the top 20 genes that have the most transcripts ORFs with annotations per transcript. The column ORF.DomainAnnot shows the PFAM ID of one of the domains overlapping the corresponding ORF. Interestingly often several ORFs cover the same domain in some of the proteins.

``` {r top annotated ORFs,echo=FALSE}
    
    head(tab[order(tab$NumORFAnnot,decreasing=T),c("geneID","OrfTransID","NumORFAnnot","ORF.DomainAnnot")],20)
    
```

## Domain analysis

Number of different Domain annotations found in Split ORFs.

``` {r plot results for annotations,echo=FALSE}
    
    length(unique(names(table(unlist(tab$ORF.DomainAnnot)))))

```


Analysis of the type of domains that are part of SplitORFs looking per gene, this avoids counting similar transcripts from the same gene.

``` {r stats annotations Gene, echo=FALSE}
    #create subset with one transcript per gene, with the highest number of ORFs that overlap an annotation
    subTab=tab[order(tab$NumORFAnnot,decreasing=T),]
    subTab=subTab[!duplicated(subTab$geneID),]
    hist(table(unlist(subTab$ORF.DomainAnnot)),xlab="Number of domain type occurrences",main="Frequency of Domain type occurrences for all unique genes",col="blue",breaks=40)
```

Show the top 50 most frequent domains found in Split Orfs (counting once per gene).
``` {r plot more results Gene,echo=FALSE}
    head(sort(table(unlist(subTab$ORF.DomainAnnot)),decreasing=T),50)

```


Analysis of the type of domains that are part of SplitORFs looking at all transcripts, this incorporates also the case that two transcripts may come from the same gene.

``` {r stats annotations, echo=FALSE}
    hist(table(unlist(tab$ORF.DomainAnnot)),xlab="Number of domain type occurrences",main="Frequency of Domain type occurrences of all transcripts",col="blue",breaks=40)
```


Show the top 50 most frequent domains found in Split Orfs (counting all transcripts).
``` {r plot more results,echo=FALSE}
    head(sort(table(unlist(tab$ORF.DomainAnnot)),decreasing=T),50)

```

This document was created with R Markdown and the Knit package. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see http://rmarkdown.rstudio.com.